clarity/src/vm/analysis/type_checker/mod.rs

```diff
@@ -307,34 +371,311 @@ impl FunctionType {
+                clarity_version,
+                1,
+                tracker,
+            )
+            .is_err()
+            {
+                return Err(CheckErrors::IncompatibleTrait(
+                    expected_trait_identifier.clone(),
+                    actual_trait_identifier.clone(),
+                )
+                .into());
+            }
+        } else {
+            return Err(CheckErrors::IncompatibleTrait(
+                expected_trait_identifier.clone(),
+                actual_trait_identifier.clone(),
+            )
+            .into());
+        }
+    }
+    Ok(())
+}
+
+fn inner_type_check_type<T: CostTracker>(
+    db: &mut AnalysisDatabase,
+    contract_context: Option<&ContractContext>,
+    actual_type: &TypeSignature,
+    expected_type: &TypeSignature,
+    clarity_version: ClarityVersion,
+    depth: u8,
+    tracker: &mut T,
+) -> TypeResult {
+    if depth > MAX_TYPE_DEPTH {
+        return Err(CheckErrors::TypeSignatureTooDeep.into());
+    }
+
+    // Recurse into values to check embedded traits properly
+    match (actual_type, expected_type) {
+        (
+            TypeSignature::OptionalType(atom_inner_type),
+            TypeSignature::OptionalType(expected_inner_type),
+        ) => {
+            inner_type_check_type(
+                db,
+                contract_context,
+                atom_inner_type,
+                expected_inner_type,
+                clarity_version,
+                depth + 1,
+                tracker,
+            )?;
+        }
+        (
+            TypeSignature::ResponseType(atom_inner_types),
+            TypeSignature::ResponseType(expected_inner_types),
+        ) => {
+            inner_type_check_type(
+                db,
+                contract_context,
+                &atom_inner_types.0,
+                &expected_inner_types.0,
+                clarity_version,
+                depth + 1,
+                tracker,
+            )?;
+            inner_type_check_type(
+                db,
+                contract_context,
+                &atom_inner_types.1,
+                &expected_inner_types.1,
+                clarity_version,
+                depth + 1,
+                tracker,
+            )?;
+        }
+        (
+            TypeSignature::SequenceType(SequenceSubtype::ListType(atom_list_type)),
+            TypeSignature::SequenceType(SequenceSubtype::ListType(expected_list_type)),
+        ) => {
+            inner_type_check_type(
+                db,
+                contract_context,
+                atom_list_type.get_list_item_type(),
+                expected_list_type.get_list_item_type(),
+                clarity_version,
+                depth + 1,
+                tracker,
+            )?;
+        }
+        (
+            TypeSignature::TupleType(atom_tuple_type),
+            TypeSignature::TupleType(expected_tuple_type),
+        ) => {
+            for (name, atom_field_type) in atom_tuple_type.get_type_map() {
```

Comment:
This verifies that the given tuple type provides only key/value pairs that are in the expected tuple type.  But it does _not_ test that all key/value pairs in the expected tuple type are in the given tuple type.  Is this intentional, or is this an oversight?

Reply:
You're right, that does look like an oversight. It should be reversed. Thanks for catching it!